### PR TITLE
Add Gradient descent algorithms

### PR DESCRIPTION
**Pull request by paquiteau**
As announced in #194 , this PR adds support for gradient descent algorithms with various gradient direction and speed preconditioning.

### CODE REVIEW

**Comment by sfarrens** [comment]
Hi @paquiteau thanks for the suggestion and the PR! I will review this properly once all the CI issues are resolved. If you are having any problem with some of the [WPS](https://github.com/wemake-services/wemake-python-styleguide) errors let me know as we can relax some errors for certain files. In particular, it will be necessary for `algorithms/__init_.py` in order to support the `*` imports.

**Comment by paquiteau** [comment]
I fixed most of the issues with linting, and updated  `setup.cfg` when I could not find another way.  I still have to write some test specific to the gradient descent though

File: modopt/opt/algorithms/gradient_descent.py [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+                or self._cost_func.get_cost(self._x_new)
+            )
+
+    def update_grad_dir(self, grad):
+        """Update the gradient descent direction."""
+        self._dir_grad = grad
+
+    def update_grad_speed(self, grad):
+        """Update the gradient descent speed."""
+        pass
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:150]: I guess here you are simply defining a method you expect the child classes to override, right? I don't really think this is needed unless the child class is forced to override this (*e.g.* as an abstract method).

File: modopt/opt/algorithms/gradient_descent.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+        """
+        metrics = {}
+        for obs in self._observers['cv_metrics']:
+            metrics[obs.name] = obs.retrieve_metrics()
+        self.metrics = metrics
+
+    def reset(self):
+        """Reset internal state of the algorithm."""
+        pass
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:189]: Same point regarding method overriding.

File: modopt/opt/algorithms/gradient_descent.py [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+        # no scale factor
+        self._speed_grad = 1.0
+        self._eps = 0
+
+
+class AdaGenericGradOpt(GenericGradOpt):
+    """Generic Grad descent Algorithm with ADA acceleration scheme."""
+
+    def update_grad_speed(self, grad):
+        """Ada Acceleration Scheme."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:209]: Is there a reference you can include for this?

File: modopt/opt/algorithms/gradient_descent.py [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+    Raises
+    ------
+    ValueError
+        If gamma is outside ]0,1[
+
+    Notes
+    -----
+    The gradient speed is defined as :
+
+    .. math:: s_k = \gamma s_{k-1}  + (1-\gamma) * |\nabla f|^2
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:230]: Is there a reference you can include for this?

File: modopt/opt/algorithms/gradient_descent.py (hunk 3 above)
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:209]: Should probably added a properties section as there is an input parameter (`grad`)

File: modopt/opt/algorithms/gradient_descent.py [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+
+    def __init__(self, *args, gamma=0.5, **kwargs):
+        super().__init__(*args, **kwargs)
+        if gamma < 0 or gamma > 1:
+            raise ValueError('gamma is outside of range [0,1]')
+        self._check_param(gamma)
+        self._gamma = gamma
+
+    def update_grad_speed(self, grad):
+        """Rmsprop update speed."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:242]: Same point re: `grad`

File: modopt/opt/algorithms/gradient_descent.py [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+    Parameters
+    ----------
+    beta: float, default 0.9
+        update weight for the momentum.
+
+    Notes
+    -----
+    The momentum update is defined as:
+
+    .. math:: m_k = \beta * m_{k-1} + \nabla f(x_k)
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:260]: Is there a reference you can include for this?

File: modopt/opt/algorithms/gradient_descent.py [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+    def __init__(self, *args, beta=0.9, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._check_param(beta)
+        self._beta = beta
+        # no scale factor
+        self._speed_grad = 1.0
+        self._eps = 0
+
+    def update_grad_dir(self, grad):
+        """Momentum gradient direction update."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:272]: Same point re: grad

File: modopt/opt/algorithms/gradient_descent.py [hunk 8]
```diff
@@ -0,0 +1,354 @@
...
+        If gamma or beta is outside ]0,1[
+
+    Notes
+    -----
+    The ADAM optimizer is defined as:
+
+    .. math::
+        m_{k+1} = \frac{1}{1-\beta^k}(\beta*m_{k}+(1-\beta)*|\nabla f_k|^2)
+    .. math::
+        s_{k+1} = \frac{1}{1-\gamma^k}(\gamma*s_k+(1-\gamma)*\nabla f_k)
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:302]: Is there a reference you can include for this?

File: modopt/opt/algorithms/gradient_descent.py [hunk 9]
```diff
@@ -0,0 +1,354 @@
...
+            raise ValueError('gamma is outside of range [0,1]')
+        if beta < 0 or beta >= 1:
+            raise ValueError('beta is outside of range [0,1]')
+        self._gamma = gamma
+        self._beta = beta
+        self._beta_pow = 1
+        self._gamma_pow = 1
+
+    def update_grad_dir(self, grad):
+        """ADAM Update of gradient direction."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:319]: Same point re: grad

File: modopt/opt/algorithms/gradient_descent.py [hunk 10]
```diff
@@ -0,0 +1,354 @@
...
+        self._beta_pow *= self._beta
+
+        self._dir_grad = (1.0 / (1.0 - self._beta_pow)) * (
+            self._beta * self._dir_grad + (1 - self._beta) * grad
+        )
+
+    def update_grad_speed(self, grad):
+        """ADAM Updatae of gradient speed."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:327]: Same point re: grad

File: modopt/opt/algorithms/gradient_descent.py [hunk 11]
```diff
@@ -0,0 +1,354 @@
...
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self._grad_memory = np.zeros(
+            (self.epoch_size, *self._x_old.shape),
+            dtype=self._x_old.dtype,
+        )
+
+    def update_grad_dir(self, grad):
+        """SAGA Update gradient direction."""
```
Comment (sfarrens) [inline modopt/opt/algorithms/gradient_descent.py:351]: Same point re: grad

File: modopt/tests/test_opt.py [hunk 12]
```diff
@@ -37,292 +37,6 @@ class Dummy(object):
     pass
```
Comment (sfarrens) [inline modopt/tests/test_opt.py:39]: excess blank line

**Comment by sfarrens** [comment]
Hi @paquiteau!

I made a first pass on your PR. Firstly, very nice and meticulous work 👏 . 

- I proposed some minor changes to `gradient_descent.py` just to clean things up a bit, but otherwise the implementation looks good and the tests pass. The main thing you could add would be some references as it is good to have a direct comparison between a given equation in a given paper/book and the code implemented.
- It appears there are some minor conflicts with the `develop` branch. This should be easy enough to fix if you pull the latest commits from `develop`.

Once these points are resolved I will make a final pass, in particular for the moved algorithms, which I believe should not have changed much.

File: modopt/opt/algorithms/gradient_descent.py (hunk 1 above)
Comment (paquiteau) [reply modopt/opt/algorithms/gradient_descent.py:150]: I had to defined this empty method for the generic case. (cf `_update` method, which is commont to all gradient descent algorithms.

File: modopt/opt/algorithms/gradient_descent.py (hunk 2 above)
Comment (paquiteau) [reply modopt/opt/algorithms/gradient_descent.py:189]: same answer ;) some class reimplements it (AdaGrad for example), other not.

**Comment by paquiteau** [comment]
Thanks for the review. 
I added the references where it was required, and documented what was not :) 

Also, I change the `update_grad_dir` and `_update_grad_speed` to private methods, these should not be changed/called by the user, as they are directly called inside `_update`.

**Comment by sfarrens** [comment]
Hi @paquiteau great! I still see a couple or merge conflicts listed between your branch and `develop`.

```
Conflicting files
modopt/opt/algorithms.py
modopt/tests/test_opt.py
```

File: modopt/opt/algorithms/gradient_descent.py [hunk 13]
```diff
@@ -0,0 +1,354 @@
...
+
+        """
+        self._run_alg(max_iter)
+
+        # retrieve metrics results
+        self.retrieve_outputs()
+
+        self.x_final = self._x_new
+
+    def _update(self):
```
[the remark this answers is not in the document]
Comment (sfarrens) [reply modopt/opt/algorithms/gradient_descent.py:138]: @paquiteau this thread still needs to be resolved

**Comment by sfarrens** [comment]
@paquiteau I have finished checking the other modules and, apart from some very minor things that I can clean up myself when I make the next release, everything looks good.

So, as soon as you resolve the last thread and we figure out the the conflict issue this PR can be merged! 👏

**Review by sfarrens — APPROVED** [review]

**Pull request merged**